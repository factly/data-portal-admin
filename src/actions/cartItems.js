import axios from '../utils/axios';
import {
  ADD_CARTITEMS,
  SET_CARTITEM_LOADING,
  SET_CARTITEM_REQUEST,
  SET_CARTITEM_IDS,
  CARTITEM_API,
} from '../constants/cartItems';
import { addProducts } from './products';
import { getIds, getValues, deleteKeys, buildObjectOfItems } from '../utils/objects';

export const loadCartItems = (id, page = 1, limit = 5) => {
  return async (dispatch, getState) => {
    const {
      cartItems: { req },
    } = getState();

    let ids;
    for (let item of req) {
      if (item.page === page && item.limit === limit) {
        ids = [...item.ids];
      }
    }

    if (ids) {
      dispatch(setCartItemIds(ids));
      return;
    }

    dispatch(setLoading(true));

    const response = await axios({
      url: `${CARTITEM_API(id)}?page=${page}&limit=${limit}`,
      method: 'get',
    });

    const { nodes, total } = response.data;
    const currentPageIds = getIds(nodes);
    const currentReq = { id, page: page, limit: limit, ids: currentPageIds };
    dispatch(setCartItemRequest(currentReq, total));
    dispatch(addCartItems(nodes));
    dispatch(setCartItemIds(currentPageIds));

    dispatch(setLoading(false));
  };
};

const setLoading = (loading) => {
  return {
    type: SET_CARTITEM_LOADING,
    payload: { loading },
  };
};

export const addCartItems = (cartItems) => (dispatch) => {
  const products = getValues(cartItems, 'product');
  dispatch(addProducts(products));

  dispatch({
    type: ADD_CARTITEMS,
    payload: {
      cartItems: buildObjectOfItems(deleteKeys(cartItems, ['product'])),
    },
  });
};

const setCartItemRequest = (req, total) => {
  return {
    type: SET_CARTITEM_REQUEST,
    payload: { req, total },
  };
};

const setCartItemIds = (ids) => {
  return {
    type: SET_CARTITEM_IDS,
    payload: { ids },
  };
};
