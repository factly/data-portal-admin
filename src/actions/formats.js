import axios from 'axios';
import {
  ADD_FORMAT,
  ADD_FORMATS,
  SET_FORMAT_LOADING,
  SET_FORMAT_REQUEST,
  SET_FORMAT_IDS,
  RESET_FORMAT,
  FORMAT_API,
} from '../constants/formats';
import { getIds, buildObjectOfItems } from '../utils/objects';

export const loadFormats = (query) => {
  return async (dispatch, getState) => {
    dispatch(setLoading(true));

    const response = await axios({
      url:FORMAT_API,
      method: 'get',
      params: query
    });

    const { nodes, total } = response.data;
    const currentPageIds = getIds(nodes);
    const currentReq = { ...query, ids: currentPageIds };
    dispatch(setFormatRequest(currentReq, total));
    dispatch(addFormats(nodes));
    dispatch(setFormatIds(currentPageIds));

    dispatch(setLoading(false));
  };
};

export const createFormat = (data) => {
  return async (dispatch, getState) => {
    dispatch(setLoading(true));

    await axios({
      url: FORMAT_API,
      method: 'post',
      data: data,
    });

    dispatch(resetFormat());
    dispatch(setLoading(false));
  };
};

export const updateFormat = (id, data) => {
  return async (dispatch, getState) => {
    let url = `${FORMAT_API}/${id}`;

    dispatch(setLoading(true));

    const response = await axios({
      url: url,
      method: 'put',
      data: data,
    });

    dispatch(addFormat(response.data));
    dispatch(setLoading(false));
  };
};

export const deleteFormat = (id) => {
  return async (dispatch, getState) => {
    let url = `${FORMAT_API}/${id}`;

    dispatch(setLoading(true));

    await axios({
      url: url,
      method: 'delete',
    });

    dispatch(resetFormat());
    dispatch(setLoading(false));
  };
};

export const getFormat = (id) => {
  return async (dispatch, getState) => {
    const {
      formats: { items },
    } = getState();

    if (items[id]) {
      return;
    }

    dispatch(setLoading(true));

    const response = await axios({
      url: `${FORMAT_API}/${id}`,
      method: 'get',
    });
    dispatch(addFormat(response.data));

    dispatch(setLoading(false));
  };
};

export const setLoading = (loading) => {
  return {
    type: SET_FORMAT_LOADING,
    payload: { loading },
  };
};

export const addFormat = (format) => {
  return {
    type: ADD_FORMAT,
    payload: { format },
  };
};

export const addFormats = (formats) => {
  return {
    type: ADD_FORMATS,
    payload: {
      formats: buildObjectOfItems(formats),
    },
  };
};

export const setFormatRequest = (req, total) => {
  return {
    type: SET_FORMAT_REQUEST,
    payload: { req, total },
  };
};

export const setFormatIds = (ids) => {
  return {
    type: SET_FORMAT_IDS,
    payload: { ids },
  };
};

export const resetFormat = () => {
  return {
    type: RESET_FORMAT,
  };
};
