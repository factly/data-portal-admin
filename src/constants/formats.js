export const ADD_FORMAT = 'ADD_FORMAT';
export const ADD_FORMATS = 'ADD_FORMATS';
export const SET_FORMAT_LOADING = 'SET_FORMAT_LOADING';
export const SET_FORMAT_REQUEST = 'SET_FORMAT_REQUEST';
export const SET_FORMAT_IDS = 'SET_FORMAT_IDS';
export const RESET_FORMAT = 'RESET_FORMAT';

export const FORMAT_API = `/formats`;
