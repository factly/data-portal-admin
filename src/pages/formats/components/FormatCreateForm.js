import React from 'react';
import { useHistory } from 'react-router-dom';
import { Form, Input, Button, Switch, notification } from 'antd';
import { CloseOutlined, CheckOutlined } from '@ant-design/icons';

const formItemLayout = {
  labelCol: { span: 6 },
  wrapperCol: { span: 14 },
};

const CurrencyCreateForm = ({ onSubmit, data = {} }) => {
  const history = useHistory();

  const onFinish = (values) => {
    onSubmit(values)
      .then(() => {
        notification.success({
          message: 'Success',
          description: 'Format added succesfully',
        });
        history.push('/formats');
      })
      .catch(() => {
        notification.error({
          message: 'Error',
          description: 'Something went wrong',
        });
      });
  };

  return (
    <Form
      name="formats_create"
      {...formItemLayout}
      initialValues={{ is_default: true, ...data }}
      onFinish={onFinish}
    >
      <Form.Item
        label="Name"
        name="name"
        rules={[
          {
            required: true,
            message: 'Please enter format name!',
          },
        ]}
      >
        <Input placeholder="Ex. PDF" />
      </Form.Item>

      <Form.Item
        label="Description"
        name="description"
        rules={[
          {
            required: true,
            message: 'Please enter format description!',
          },
        ]}
      >
        <Input placeholder="Ex. Locked file" />
      </Form.Item>

      <Form.Item
        label="Is Default"
        name="is_default"
        valuePropName="checked"
        rules={[
          {
            required: true,
            message: 'Please select an option!',
          },
        ]}
      >
        <Switch
          checkedChildren={<CheckOutlined />}
          unCheckedChildren={<CloseOutlined />}
          defaultChecked
        />
      </Form.Item>

      <Form.Item wrapperCol={{ span: 12, offset: 6 }}>
        <Button type="primary" htmlType="submit">
          Submit
        </Button>
      </Form.Item>
    </Form>
  );
};

export default CurrencyCreateForm;
