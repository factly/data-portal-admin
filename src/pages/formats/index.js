import React from 'react';
import FormatsList from './components/FormatsList';
import { Button } from 'antd';
import { Link } from 'react-router-dom';

const Formats = () => {
  return (
    <>
      <Link to={'/formats/create'}>
        <Button type="primary" style={{ marginBottom: 16 }}>
          Add Format
        </Button>
      </Link>
      <FormatsList />
    </>
  );
};

export default Formats;
