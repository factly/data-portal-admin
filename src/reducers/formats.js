import produce from 'immer';
import {
  ADD_FORMAT,
  ADD_FORMATS,
  SET_FORMAT_LOADING,
  SET_FORMAT_REQUEST,
  SET_FORMAT_IDS,
  RESET_FORMAT,
} from '../constants/formats';

const initialState = {
  loading: true,
  ids: [],
  req: [],
  items: {},
  total: 0,
};

const formatsReducer = produce((draft, action = {}) => {
  switch (action.type) {
    case SET_FORMAT_LOADING:
      draft.loading = action.payload.loading;
      return;
    case ADD_FORMAT: {
      const { format } = action.payload;
      draft.items[format.id] = format;
      return;
    }
    case ADD_FORMATS: {
      const { formats } = action.payload;
      Object.assign(draft.items, formats);
      return;
    }
    case SET_FORMAT_IDS:
      draft.ids = action.payload.ids;
      return;
    case SET_FORMAT_REQUEST: {
      const { req, total } = action.payload;
      draft.req.push(req);
      draft.total = total;
      return;
    }
    case RESET_FORMAT:
      return initialState;
  }
}, initialState);

export default formatsReducer;
