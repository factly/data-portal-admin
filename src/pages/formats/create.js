import React from 'react';
import { useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';

import FormatCreateForm from './components/FormatCreateForm';
import { createFormat } from '../../actions/formats';

const FormatCreate = () => {
  const history = useHistory();
  const dispatch = useDispatch();

  const onCreate = (values) => {
    dispatch(createFormat(values));
    history.push('/formats');
  };

  return <FormatCreateForm onSubmit={onCreate} />;
};

export default FormatCreate;
