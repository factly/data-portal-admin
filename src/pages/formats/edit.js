import React from 'react';
import { useParams, useHistory } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { Skeleton } from 'antd';

import FormatCreateForm from './components/FormatCreateForm';
import { getFormat, updateFormat } from '../../actions/formats';

const EditFormat = () => {
  const { id } = useParams();
  const history = useHistory();
  const dispatch = useDispatch();
  const { format } = useSelector(({ formats }) => formats);

  React.useEffect(() => {
    dispatch(getFormat(id));
  }, []);

  if (!format.id) return <Skeleton />;

  const onUpdate = async (values) => {
    await dispatch(updateFormat(id, values));
    history.push('/formats');
  };

  return <FormatCreateForm data={format} onSubmit={onUpdate} />;
};

export default EditFormat;
