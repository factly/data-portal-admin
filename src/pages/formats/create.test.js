import React from 'react';
import { useHistory } from 'react-router-dom';
import { useDispatch, Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount, shallow } from 'enzyme';
import { act } from '@testing-library/react';

import '../../matchMedia.mock';
import CreateFormat from './create';
import * as actions from '../../actions/formats';
import FormatCreateForm from './components/FormatCreateForm';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(),
}));

jest.mock('../../actions/formats', () => ({
  createFormat: jest.fn(),
}));

describe('Formats create component', () => {
  let store;
  let mockedDispatch;

  store = mockStore({
    formats: {
      loading: false,
      ids: [],
      req: [],
      items: {},
      total: 0,
    },
  });
  store.dispatch = jest.fn(() => ({}));
  mockedDispatch = jest.fn(() => Promise.resolve({}));
  useDispatch.mockReturnValue(mockedDispatch);
  describe('snapshot testing', () => {
    it('should render the component', () => {
      const component = shallow(<CreateFormat />);
      expect(component).toMatchSnapshot();
    });
  });
  describe('component testing', () => {
    let wrapper;
    afterEach(() => {
      wrapper.unmount();
    });
    it('should call createFormat', (done) => {
      actions.createFormat.mockReset();
      const push = jest.fn();
      useHistory.mockReturnValueOnce({ push });
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <CreateFormat />
          </Provider>,
        );
      });
      wrapper.find(FormatCreateForm).props().onSubmit({ test: 'test' });
      setTimeout(() => {
        expect(actions.createFormat).toHaveBeenCalledWith({ test: 'test' });
        expect(push).toHaveBeenCalledWith('/formats');
        done();
      }, 0);
    });
  });
});
