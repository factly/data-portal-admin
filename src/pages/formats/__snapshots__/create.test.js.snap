// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Formats create component snapshot testing should render the component 1`] = `
<CurrencyCreateForm
  onSubmit={[Function]}
/>
`;
